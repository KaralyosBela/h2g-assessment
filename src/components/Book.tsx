import { IBook } from "../interface/book.inteface";
import classes from "../components/Book.module.css";
import { useNavigate } from "react-router-dom";
import { useDispatch } from "react-redux";
import { AppDispatch } from "../app/store";
import { editBook } from "../features/bookSlice";
import {RiEditCircleFill} from "react-icons/ri"

interface Props {
  book: IBook;
}

export const Book: React.FC<Props> = ({ book }) => {
  const navigate = useNavigate();
  const dispatch = useDispatch<AppDispatch>();

  const navigateToEditBook = (id: number) => {
    navigate(`/books/edit/${id}`);
  };

  const changeBookStatus = (status: string) => {
    if (status === "reading" && book.reading === false) {
      dispatch(
        editBook({
          ...book,
          reading: !book.reading,
          read: !book.read,
        })
      );
    } else if (status === "read" && book.read === false) {
      dispatch(
        editBook({
          ...book,
          reading: !book.reading,
          read: !book.read,
        })
      );
    }
  };

  return (
    <div className={classes.bookElement}>
      {book.title}
      <div className={classes.interaction}>
        <button data-testid="edit-button" className={classes.editBtn} onClick={() => navigateToEditBook(book.id)}>Edit<RiEditCircleFill size={10} color={"white"}/></button>
        <button data-testid="new-button">New</button>
        <button
          data-testid="reading-button"
          className={book.reading ? classes.active : ""}
          onClick={() => changeBookStatus("reading")}>Reading</button>
        <button
          data-testid="read-button"
          className={book.read ? classes.active : ""}
          onClick={() => changeBookStatus("read")}>Read
        </button>
      </div>
    </div>
  );
};
