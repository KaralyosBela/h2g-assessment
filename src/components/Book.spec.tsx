import { fireEvent, render, screen } from "@testing-library/react";
import { Book } from "./Book";

// const navigate = useNavigate();
//   const dispatch = useDispatch<AppDispatch>();

const mockDispatch = jest.fn();
const mockEditBook = jest.fn();
const mockUseDispatch = jest.fn(() => mockDispatch);
const mockNavigate = jest.fn();
const mockUseNavigate = jest.fn(() => ({ mockNavigate }));

jest
  .mock("react-router-dom", () => ({
    useNavigate: () => mockUseNavigate,
  }))
  .mock("react-redux", () => ({
    useDispatch: () => mockUseDispatch,
  }))
  .mock("../features/bookSlice", () => ({
    editBook: (...params: unknown[]) => mockEditBook(params),
  }));

describe("book", () => {
  it("should render component when is read", () => {
    const book = {
      id: 1,
      title: "testTitle",
      author: "testAuthor",
      description: "testDescription",
      reading: false,
      read: true,
    };
    const { container } = render(<Book book={book} />);
    expect(container).toMatchSnapshot();
  });

  it("should render component when is reading", () => {
    const book = {
      id: 1,
      title: "testTitle",
      author: "testAuthor",
      description: "testDescription",
      reading: true,
      read: false,
    };
    const { container } = render(<Book book={book} />);
    expect(container).toMatchSnapshot();
  });

  it("should render component when edit is clicked", () => {
    const book = {
      id: 1,
      title: "testTitle",
      author: "testAuthor",
      description: "testDescription",
      reading: true,
      read: false,
    };
    const { container } = render(<Book book={book} />);
    fireEvent.click(screen.getByTestId("edit-button"));
    expect(mockUseNavigate.mock.calls).toEqual([["/books/edit/1"]]);
    expect(container).toMatchSnapshot();
  });

  it("should render component when reading is clicked", () => {
    const book = {
      id: 1,
      title: "testTitle",
      author: "testAuthor",
      description: "testDescription",
      reading: false,
      read: true,
    };
    mockEditBook.mockReturnValue("mockEditBook");
    const { container } = render(<Book book={book} />);
    fireEvent.click(screen.getByTestId("reading-button"));
    expect(mockUseDispatch.mock.calls).toEqual([["mockEditBook"]]);
    expect(mockEditBook.mock.calls).toEqual([
      [
        [
          {
            author: "testAuthor",
            description: "testDescription",
            id: 1,
            read: false,
            reading: true,
            title: "testTitle",
          },
        ],
      ],
    ]);
    expect(container).toMatchSnapshot();
  });

  it("should render component when read is clicked", () => {
    const book = {
      id: 1,
      title: "testTitle",
      author: "testAuthor",
      description: "testDescription",
      reading: true,
      read: false,
    };
    mockEditBook.mockReturnValue("mockEditBook");
    const { container } = render(<Book book={book} />);
    fireEvent.click(screen.getByTestId("read-button"));
    expect(mockUseDispatch.mock.calls).toEqual([["mockEditBook"]]);
    expect(mockEditBook.mock.calls).toEqual([
      [
        [
          {
            author: "testAuthor",
            description: "testDescription",
            id: 1,
            read: true,
            reading: false,
            title: "testTitle",
          },
        ],
      ],
    ]);
    expect(container).toMatchSnapshot();
  });
});
