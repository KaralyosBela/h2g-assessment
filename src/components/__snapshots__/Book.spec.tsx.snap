// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`book should render component when edit is clicked 1`] = `
<div>
  <div
    class="bookElement"
  >
    testTitle
    <div
      class="interaction"
    >
      <button
        data-testid="edit-button"
      >
        edit
      </button>
      <button
        class="active"
        data-testid="reading-button"
      >
        reading
      </button>
      <button
        class=""
        data-testid="read-button"
      >
        read
      </button>
    </div>
  </div>
</div>
`;

exports[`book should render component when is read 1`] = `
<div>
  <div
    class="bookElement"
  >
    testTitle
    <div
      class="interaction"
    >
      <button
        data-testid="edit-button"
      >
        edit
      </button>
      <button
        class=""
        data-testid="reading-button"
      >
        reading
      </button>
      <button
        class="active"
        data-testid="read-button"
      >
        read
      </button>
    </div>
  </div>
</div>
`;

exports[`book should render component when is reading 1`] = `
<div>
  <div
    class="bookElement"
  >
    testTitle
    <div
      class="interaction"
    >
      <button
        data-testid="edit-button"
      >
        edit
      </button>
      <button
        class="active"
        data-testid="reading-button"
      >
        reading
      </button>
      <button
        class=""
        data-testid="read-button"
      >
        read
      </button>
    </div>
  </div>
</div>
`;

exports[`book should render component when read is clicked 1`] = `
<div>
  <div
    class="bookElement"
  >
    testTitle
    <div
      class="interaction"
    >
      <button
        data-testid="edit-button"
      >
        edit
      </button>
      <button
        class="active"
        data-testid="reading-button"
      >
        reading
      </button>
      <button
        class=""
        data-testid="read-button"
      >
        read
      </button>
    </div>
  </div>
</div>
`;

exports[`book should render component when reading is clicked 1`] = `
<div>
  <div
    class="bookElement"
  >
    testTitle
    <div
      class="interaction"
    >
      <button
        data-testid="edit-button"
      >
        edit
      </button>
      <button
        class=""
        data-testid="reading-button"
      >
        reading
      </button>
      <button
        class="active"
        data-testid="read-button"
      >
        read
      </button>
    </div>
  </div>
</div>
`;
